Added layer tint color

Gnumaru opened the pull request:
Initial implementation for tinting a tile layer. the matter of stacking the tint of group layers in the same manner it happens with opacity in the current version of tiled was not handled. Also, tinting non tile layers was also not handled.

Closes #2683

bjorn commented on src/libtiled/maprenderer.cpp:
```diff
@@ -335,7 +350,17 @@ void CellRenderer::render(const Cell &cell, const QPointF &pos, const QSizeF &si
     mPainter->setTransform(transform);
-    mPainter->drawPixmap(target, image, source);
+    QPixmap tintedimg = image;
+    if(layer){
```
When is layer not a valid pointer? Could it be turned into a reference?

Gnumaru replied:
Now I'm storing the tintcolor in the cellrenderer as you said, so passing around a pointer or reference to the layer becomes unnecessary.

bjorn commented on src/libtiled/maprenderer.cpp:
```diff
@@ -335,7 +350,17 @@ void CellRenderer::render(const Cell &cell, const QPointF &pos, const QSizeF &si
+    if(layer){
+        QColor tintColor = layer->tintColor();
+        if(tintColor != QColor(255, 255, 255, 255)){
+            tintedimg = QPixmap::fromImage(tinted(QImage(image.toImage())
```
Instead of passing `QImage(image.toImage())` here, please take `QPixmap` in the `tinted` function and use `drawPixmap` there instead of `drawImage`. I doubt it matters for performance, but it's less code.

Gnumaru replied:
done

bjorn commented on src/libtiled/maprenderer.cpp:
```diff
@@ -335,7 +350,17 @@ void CellRenderer::render(const Cell &cell, const QPointF &pos, const QSizeF &si
+        QColor tintColor = layer->tintColor();
+        if(tintColor != QColor(255, 255, 255, 255)){
+            tintedimg = QPixmap::fromImage(tinted(QImage(image.toImage())
+                                                    ,layer->tintColor()
```
Either use the local `tileColor` here or use `layer->tintColor` also in the condition. I would probably remove the local `tintColor`.

Gnumaru replied:
done

bjorn commented on src/libtiled/maprenderer.cpp:
```diff
@@ -259,7 +274,7 @@ void CellRenderer::render(const Cell &cell, const QPointF &pos, const QSizeF &si
     // The USHRT_MAX limit is rather arbitrary but avoids a crash in
     // drawPixmapFragments for a large number of fragments.
     if (mTile != tile || mFragments.size() == USHRT_MAX)
```
In theory we need to flush as well when the layer, or at least the tint color, changed. In practice I don't think a `CellRenderer` instance is ever used with different layers so the tint color won't change. As such, I would actually suggest we add the tint color as constructor parameter to the `CellRenderer` instead.

Gnumaru replied:
done

bjorn commented on src/tiled/changelayer.cpp:
```diff
@@ -121,6 +121,29 @@ SetLayerOpacity::SetLayerOpacity(Document *document,
+
+SetLayerTintColor::SetLayerTintColor(Document *document,
+                                 Layer *layer,
+                                 QColor tintColor)
```
Please place the constructor before the methods.

Gnumaru replied:
done

bjorn commented on src/tiled/propertybrowser.cpp:
```diff
@@ -1586,6 +1590,8 @@ void PropertyBrowser::addProperties()
     if (QtProperty *fontProperty = mIdToProperty.value(FontProperty))
         setExpanded(items(fontProperty).constFirst(), false);
+    if (QtProperty *tintColorProperty = mIdToProperty.value(TintColorProperty))
+                setExpanded(items(tintColorProperty).constFirst(), false);
```
```suggestion
        setExpanded(items(tintColorProperty).constFirst(), false);
```

bjorn commented on src/libtiled/maprenderer.cpp:
```diff
@@ -48,6 +48,21 @@
+    QImage resultImage(image.size(), QImage::Format_ARGB32_Premultiplied);
+    QPainter painter(&resultImage);
+    painter.drawImage(0, 0, image);
+    painter.setCompositionMode(mode);
```
Since I don't think we're planning to support other composition modes, please just put `QPainter::CompositionMode_Multiply` here.

Gnumaru commented on src/libtiled/maprenderer.cpp:
```diff
@@ -48,6 +48,21 @@
+    painter.setCompositionMode(mode);
+    painter.fillRect(resultImage.rect(), color);
+    painter.end();
+    resultImage.setAlphaChannel(image.alphaChannel());
```
done

bjorn commented on src/libtiled/isometricrenderer.cpp:
```diff
@@ -401,6 +401,7 @@ void IsometricRenderer::drawMapObject(QPainter *painter,
             painter->setPen(pen);
             painter->drawRect(rect);
             pen.setStyle(Qt::DotLine);
+            // since now layers will have a tint color, should we remove the object layer color property for objects and use the layer tintcolor here? or since object layers already have a color property, shoul we stick with the name Color and rename tintColor to color? aniways the layer color would have to be provided by the caller of IsometricRenderer::drawMapObject
```
It makes sense to keep tint color separate from this color since they do entirely different things.

I believe the layer tint color can be obtained through the object using `object->objectGroup()->tintColor()`, as per previous suggested change.

Gnumaru replied:
done

bjorn commented on src/tiled/changelayer.cpp:
```diff
@@ -138,7 +166,6 @@ void SetLayerOpacity::setOpacity(qreal opacity)
     emit mDocument->changed(LayerChangeEvent(mLayer, LayerChangeEvent::OpacityProperty));
 }
 
-
```
There are two empty lines here because the definitions above and below and for different classes. Please keep it that way. :-)

Gnumaru replied:
done

bjorn commented on src/libtiled/orthogonalrenderer.cpp:
```diff
@@ -311,7 +311,7 @@ void OrthogonalRenderer::drawTileLayer(QPainter *painter,
     painter->translate(layerPos);
 
-    CellRenderer renderer(painter, this);
+    CellRenderer renderer(painter, this, layer->tintColor(), CellRenderer::CellType::OrthogonalCells);
```
No need to add `CellRenderer::CellType::OrthogonalCells` here since that the default value.

Gnumaru commented on src/libtiled/maprenderer.cpp:
```diff
@@ -335,7 +350,17 @@ void CellRenderer::render(const Cell &cell, const QPointF &pos, const QSizeF &si
     mPainter->setTransform(transform);
-    mPainter->drawPixmap(target, image, source);
+
+    if(mTintColor != QColor(255, 255, 255, 255)){
```
done

Gnumaru commented on src/libtiled/orthogonalrenderer.cpp:
```diff
@@ -391,7 +391,7 @@ void OrthogonalRenderer::drawMapObject(QPainter *painter,
     if (!cell.isEmpty()) {
         const QSizeF size = object->size();
-        CellRenderer(painter, this).render(cell, QPointF(), size,
+        CellRenderer(painter, this, QColor(255, 255, 255, 255)).render(cell, QPointF(), size,
```
done

Gnumaru commented on src/libtiled/isometricrenderer.cpp:
```diff
@@ -372,7 +372,7 @@ void IsometricRenderer::drawMapObject(QPainter *painter,
         const QPointF pos = pixelToScreenCoords(object->position());
 
-        CellRenderer(painter, this).render(cell, pos, size,
+        CellRenderer(painter, this, QColor(255, 255, 255, 255)).render(cell, pos, size,
```
done

bjorn commented on src/tiled/changelayer.cpp:
```diff
@@ -108,6 +108,7 @@ void SetLayerLocked::swap()
     emit mDocument->changed(LayerChangeEvent(mLayer, LayerChangeEvent::LockedProperty));
 }
 
+
```
Ha, thank you for caring about consistency!

Gnumaru replied:
=)

bjorn commented:
That's really annoying that you had to do this trick with the alpha layer... I wonder why [the multiply composition mode is implemented like that](https://code.woboq.org/qt5/qtbase/src/gui/painting/qcompositionfunctions.cpp.html#_ZL23comp_func_Multiply_implPjPKjiRKT_) rather than just multiplying the alpha channel as well. I'm having trouble thinking of a use-case for this "mixing" of alpha channels:
```c++
static inline int mix_alpha(int da, int sa)
{
    return 255 - ((255 - sa) * (255 - da) >> 8);
}
```
Though I'm probably missing something just because it's not what we are looking for here.

In any case, I've pushed some simplifications, also making sure to ignore the tint color when it is not specified (in which case it is now "invalid" rather than white). This makes the "reset" button in the Properties view trivially work and could simplify the map reader and writer a little, which incidentally is I think the only part still missing in this change (and adjusting the documentation of course).

Gnumaru commented:
I have just updated my fork with the recent upstream changes.

Gnumaru commented:
I would say yes to both questions. Image layers are useful to implement static or slow moving, non tiled, backgrounds and tinting them has the same uses as tinting a conventional tile layer. On the other hand, separating in different layers something that in-game seems to be the same thing makes sense for things like fake walls where the character can enter behind the wall in a specific point and then part of the wall becomes transparent, so we can see the secrets behind. In this case, if one wants to tint both tile layers it would be easier to put them in a group and tint the group instead of having to copy the tint color from one layer to the other.

bjorn commented:
@Gnumaru Would you be interested in implementing those changes as part of this pull request?

Gnumaru commented:
I'll give it a try.

Gnumaru commented:
I implemented both the tint color inheritance on groups and the tinting of image layers. Fortunately the changes I had to do were little. I also squashed the commits but only now I realized that you should have reviewed the changes before I squashed them, I'm sorry.

bjorn commented:
@Gnumaru I assumed that commit 8575836212de8bb63aee524ad6322b9da40fc817 included all your changes and that the squashing didn't include additional changes, and have restored your master branch to that commit (and added a commit message). :-)

Gnumaru commented:
You assumed right, there were no changes after commit 8575836.

bjorn commented:
Thank you for this nice addition!

Gnumaru commented:
I'm really pleased to be able to contribute to Tiled =)

bjorn commented:
@Gnumaru https://itch.io/post/1144498 :-)
